File: SWEA/D2/1974. 스도쿠 검증/스도쿠 검증.py
# 가로를 쭉 돌면서 검사
def checking_row(row):
    cnt = 0
    for m in range(9):
        cnt += arr[row][m]
    # 한 줄의 합이 45가 아니라면 중복되는 게 있으니깐 안되는 거
    if cnt != 45:
        temp = 0
    else:
        temp = 1
    return temp


# 세로를 쭉 돌면서 검사
def checking_column(column):
    cnt = 0
    for m in range(9):
        cnt += arr[m][column]
    if cnt != 45:
        temp = 0
    else:
        temp = 1
    return temp


# 3 * 3을 돌면서 검사
def checking3by3(row, column):
    cnt = 0
    # 3으로 나눠서 어디부터 탐색할지 정해놓는다.
    x = row // 3
    y = column // 3
    for m in range(3 * x, 3 * (x+1)):
        for k in range(3 * y, 3 * (y+1)):
            cnt += arr[m][k]
    if cnt != 45:
        temp = 0
    else:
        temp = 1
    return temp


T = int(input())
for tc in range(1, T+1):
    arr = [list(map(int, input().split())) for _ in range(9)]
    answer = 1
    for i in range(9):
        for j in range(9):
            # 만약에 어디 한 군데라도 아닌 곳이 나오면 바로 멈춘다. 어차피 안되니까
            answer = checking_row(i)
            if answer != 1:
                break
            answer = checking_column(j)
            if answer != 1:
                break
            answer = checking3by3(i, j)
            if answer != 1:
                break
        if answer != 1:
            break
    print(f"#{tc} {answer}")

File: SWEA/D2/1974. 스도쿠 검증/스도쿠 검증.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Scanner;
import java.util.Set;

public class Solution {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int T = sc.nextInt();
		for (int tc = 1; tc <= T; tc++) {
			int[][] arr = new int[9][9];
			int answer = 1;
			for (int i = 0; i<9;i++) {
				for (int j = 0; j<9;j++) {
					arr[i][j] = sc.nextInt();
				}
			}
			for (int i = 0; i<9;i++) {
				for (int j = 0; j<9;j++) {
					answer = checking_row(arr, i);
					if (answer != 1) {
						break;
					}
					answer = checking_column(arr, j);
					if (answer != 1) {
						break;
					}
					answer = checking3by3(arr, i, j);
					if (answer != 1) {
						break;
					}
				}
				if (answer != 1) {
					break;
				}
			}
			System.out.println("#" + tc + " " + answer);
		}
		sc.close();
	}

	// 가로 검사
	static int checking3by3(int[][] arr, int row, int column) {
		List<Integer> check = new ArrayList<>();
		int x = row / 3;
		int y = column / 3;
		for (int m = 3 * x; m < 3 * (x + 1); m++) {
			for (int k = 3 * y; k < 3 * (y + 1); k++) {
				check.add(arr[m][k]);
			}
		}
		Set<Integer> Setcheck = new HashSet<>(check);
		if (Setcheck.size() != 9) {
			return 0;
		}
		return 1;
	}

	// 세로 검사
	static int checking_column(int[][] arr, int column) {
		List<Integer> check = new ArrayList<>();
		for (int m = 0; m < 9; m++) {
			check.add(arr[m][column]);
		}
		Set<Integer> Setcheck = new HashSet<>(check);
		if (Setcheck.size() != 9) {
			return 0;
		}
		return 1;
	}

	static int checking_row(int[][] arr, int row) {
		List<Integer> check = new ArrayList<>();
		for (int m = 0; m < 9; m++) {
			check.add(arr[row][m]);
		}
		Set<Integer> Setcheck = new HashSet<>(check);
		if (Setcheck.size() != 9) {
			return 0;
		}
		return 1;
	}
}
